### DISCUSSION TITLE
Can I do bulk inserts?

### DISCUSSION THREAD

**Discussion by brunowilliang** [pocketbase/pocketbase#3129] [Q&A]
I have almost 200,000 items in json, would I be able to convert them to apply to pocketbase?

**Comment by ganigeorgiev** [comment]
The records web APIs currently don't support bulk operations.

You can load your JSON programatically via a JS app hook (eg. in a custom console command or route action). Here is an example with a custom console command (**I haven't tested it!**):

```js
// pb_hooks/main.pb.js

// register a new "./pocketbase import /path/to/data.json" command
$app.rootCmd.addCommand(new Command({
    use: "import",
    run: (cmd, args) => {
        // read the path to the json file from the arguments list
        // aka. require("/path/to/data.json")
        const rawData = require(args[0]);

        const collection = $app.dao().findCollectionByNameOrId("yourCollection");

        $app.dao().runInTransaction((txDao) => {
            for (let item of rawData) {
                const record = new Record(collection, {
                    fieldA: item.fieldA,
                    fieldB: item.fieldB,
                    // ...
                });

                txDao.saveRecord(record);
            }
        });
    },
}));
```

More info about the JS app hooks you can find in https://pocketbase.io/docs/js-overview/.

**Reply by brunowilliang** [reply]
Got it, I'll try it this way...
I'm converting with node and using prism to generate a sqlite with the data, it would be wonderful if it allowed bulk inserts, is there any future plans to allow bulk inserts?

**Comment by insidesmart** [comment]
@ganigeorgiev I was just testing the sample code you have given.

 I think the following line is throwing some errors.

`        const rawData = require(args[0]);`

```
panic: GoError: Invalid module at github.com/dop251/goja_nodejs/require.(*RequireModule).require-fm (native)

goroutine 56 [running]:
github.com/dop251/goja.(*Runtime).toReflectValue.(*Runtime).wrapJSFunc.func1({0xc00134e090, 0x2, 0x4?})
        /home/runner/go/pkg/mod/github.com/dop251/goja@v0.0.0-20230707174833-636fdf960de1/runtime.go:2308 +0x5d4
github.com/spf13/cobra.(*Command).execute(0xc0002b4600, {0xc000b9e190, 0x1, 0x1})
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x863
github.com/spf13/cobra.(*Command).ExecuteC(0xc000005800)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(...)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/pocketbase/pocketbase.(*PocketBase).Execute.func2()
        /home/runner/work/pocketbase/pocketbase/pocketbase.go:164 +0x25
created by github.com/pocketbase/pocketbase.(*PocketBase).Execute in goroutine 1
        /home/runner/work/pocketbase/pocketbase/pocketbase.go:163 +0xe9
```

May I know what is the right way to read file content here ?

**Reply by ganigeorgiev** [reply]
Did you provide a valid path to the file to import when running the command?

If you've executed `./pocketbase import /path/to/data.json`, `args[0]` should be the string `/path/to/data.json`.
Also keep in mind that it needs to be a valid `json` or CJS module (aka. to have `module.exports = {...}`).

**Reply by insidesmart** [reply]
Thanks. It turns out the problem was file path. I was providing a relative file path as input. 

I changed the input as absolute file path. Then it worked. 

Sorry for the false alarm.

**Reply by ganigeorgiev** [reply]
Relative file path should also work but there are some caveats - see https://pocketbase.io/docs/js-overview/#relative-paths.

**Reply by insidesmart** [reply]
Thanks for the link & the quick help. Totally missed this section in the document.

**Reply by brunowilliang** [reply]
@insidesmart can you share the code with me? I don't know where to start to use pocketbase as a framework

**Reply by insidesmart** [reply]
@Brunowilliang I have explained the implementation with some sample scenario below.

- create the following in pocketbase
```
collection name = plants
fields = ["plant_name", "plant_color"]
```

- create the following content in input json file (/home/username/Desktop/input_json_plants.json)
```
[
    {"plant_name": "neem", "plant_color": "green"},
    {"plant_name": "lotus", "plant_color": "pink"}
]
```

- create the following content in hooks file at pb_hooks directory (/pocketbase_dir/pb_hooks/import_data_plants.pb.js)
```
$app.rootCmd.addCommand(new Command({
    use: "import_data_plants",
    run: (cmd, args) => {
        
        // input_json_data file path is input parameter 1
        const input_json_data = require(args[0])
        // console.log("input_json_data", JSON.stringify(input_json_data))

        const collection = $app.dao().findCollectionByNameOrId("plants");

        $app.dao().runInTransaction((txDao) => {
            // loop through all items of input_json_data
            for (let input_data of input_json_data) {
                const collection_record = new Record(collection, {
                    plant_name: input_data.plant_name,
                    plant_color: input_data.plant_color,
                });
                // console.log("collection_record", JSON.stringify(collection_record))
                txDao.saveRecord(collection_record);
            }
        });
    }
}));
```
- run the following command to import content from input_json_plants.json to plants collection in pocketbase
- format - ./pocketbase import_data_plants absolute filepath
```
./pocketbase import_data_plants "/home/username/Desktop/input_json_plants.json"
```

Let me know if you need any clarifications.

**Comment by brunowilliang** [comment]
is that I don't know how to start it as a framework, can you put the code in a git? just for me to see?

**Reply by ganigeorgiev** [reply]
Please check the guides in https://pocketbase.io/docs/js-overview/.

You have to to create `pb_hooks/something.pb.js` file(s) and bind to the specific hooks.

**Reply by brunowilliang** [reply]
@ganigeorgiev Do I have to download the source code? or i can use ./pocketbase server
I'm novice in backend and also in go, I'm trying to find a solution to see what it would be like to work as a framework, the problem is that I don't know where to start...

**Reply by ganigeorgiev** [reply]
@Brunowilliang You need only the PocketBase executable (v0.17.0+) and to create `pb_hooks/*.pb.js` file(s).

**Please again make sure to read https://pocketbase.io/docs/js-overview/.**

**Reply by brunowilliang** [reply]
@ganigeorgiev 
my structure is this:
- pb_data
- pb_hooks
  - movies.pb.js
- pb_migrations
- movies.json
- pocketbase

my movies.pb.js is this:
```js
$app.rootCmd.addCommand(new Command({
    use: "movies",
    run: (cmd, args) => {
        
      const data_file = require(args[0]);
      
      const data_json = data_file.data;
      console.log("movies", JSON.stringify(data_json));
  
      const collection = $app.dao().findCollectionByNameOrId("movies");
  
      $app.dao().runInTransaction((txDao) => {
          // loop through all items of data_json
          for (let input_data of data_json) {
              const collection_record = new Record(collection, {
                  title: input_data.title,
                  logo: input_data.logo,
                  group: input_data.group,
                  url: input_data.url
              });
              console.log("collection_record", JSON.stringify(collection_record));
              txDao.saveRecord(collection_record);
          }
      });
    }
  }));
``` 

exemple my movies.json:
```json
{
  "data": [
    {
      "title": "1",
      "logo": "https://teste.com/1",
      "group": "MOVIE 1",
      "url": "http://teste.com/1"
    },
    {
      "title": "2",
      "logo": "https://teste.com/2",
      "group": "MOVIE 2",
      "url": "http://teste.com/2"
    },
    {
      "title": "3",
      "logo": "https://teste.com/3",
      "group": "MOVIE 3",
      "url": "http://teste.com/3"
    },
    {...},
    {...},
    {...},
    {...},
    {...},
    {...},
  ],
}
```


How could I run this code?

**Reply by ganigeorgiev** [reply]
`./pocketbase movies ./movies.json`

**Reply by brunowilliang** [reply]
@ganigeorgiev 
omg it worked....

can i make connections?
in my json there is a group, I wanted it to have a relationship with a groups table

**Reply by ganigeorgiev** [reply]
> can i make connections? in my json there is a group, I wanted it to have a relationship with a groups table

Yes, if you want to save relations you'll need the id of the relation record you want to link.

To find a related "group" record, you can call
```js
const group = txDao.findFirstRecordByData("groups", "someField", someGroupValue)
console.log(group.id)
```

For other record retrieval methods, please refer to https://pocketbase.io/docs/js-records/#fetch-single-record.

**Reply by brunowilliang** [reply]
@ganigeorgiev 
OK, I think I made a mistake here because I started to play and filled with non-existent data...
do i have any command where i run to clean all my collection?

**Reply by ganigeorgiev** [reply]
> do i have any command where i run to clean all my collection?

Not out of the box, but you can use a similar approach to the above and register a new console command:

```js
$app.rootCmd.addCommand(new Command({
  use: "truncate",
  run: (cmd, args) => {
    const name = args[0];

    $app.dao().db()
        .newQuery("DELETE FROM {{" + name + "}}")
        .execute()
  }
}));
```
To run it, execute `./pocketbase truncate yourCollectionName`.

Alternatively, you can also use the sqlite3 cli tool:
```sh
sqlite3 ./pb_data/data.db 'DELETE FROM yourCollection'
```

**Reply by brunowilliang** [reply]
perfect, everything worked...
thank you very much, I will make some changes and I will try to deploy

**Reply by brunowilliang** [reply]
@ganigeorgiev
Can you please provide me with the docker to deploy on the fly?
how would I run the hooks in production?

**Reply by ganigeorgiev** [reply]
It is the same as the `pb_migrations`, aka. you can `COPY` your local `pb_hooks` inside the image:

```dockerfile
COPY ./pb_hooks /pb/pb_hooks
```

(_check the commented lines in the example Dockerfile in https://github.com/pocketbase/pocketbase/discussions/537_)

**Reply by brunowilliang** [reply]
@ganigeorgiev and how would I run this hook inside the app?

**Reply by ganigeorgiev** [reply]
_Side-note: if you want to execute custom console command on the Fly instance you'll have to SSH into your app, aka. `fly ssh console` and then run it similarly via `./pb/pocketbase ...` (see also https://fly.io/docs/flyctl/ssh/)._

**Reply by brunowilliang** [reply]
@ganigeorgiev 
it deploys, but when I send my pb_data using ssh it gets overloaded because my json contains a lot of things, just the collection is about 20 thousand items
